### PR TITLE
Added types to Index submodule

### PR DESCRIPTION
**Pull request by Yobmod**
Added types to Index submodule,
plus change some types in othe files to make them agree

### CODE REVIEW

**Review by d066zkie — APPROVED** [review]

**Review by d066zkie — APPROVED** [review]

**Review by d066zkie — APPROVED** [review]

**Comment by Byron** [comment]
@Yobmod Thanks a lot! Do you know how it's possible that somebody I didn't assign can approve changes and do so repeatedly? Is that a GitHub bug? Should I be concerned?

**Comment by Yobmod** [comment]
Oh, don't know how that is possible (or who the user is).

It's never happened until I used the draft feature. Maybe something to do with that?

File: git/index/base.py [hunk 1]
```diff
@@ -446,20 +473,21 @@ def iter_blobs(self, predicate=lambda t: True):
...
         # END for each entry
 
-    def unmerged_blobs(self):
+    def unmerged_blobs(self) -> Dict[PathLike, List[Tuple[StageType, Blob]]]:
         """
         :return:
             Iterator yielding dict(path : list( tuple( stage, Blob, ...))), being
             a dictionary associating a path in the index with a list containing
             sorted stage/blob pairs
+            ##### Does it return iterator? or just the Dict?
```
Comment (Byron) [inline git/index/base.py:482]: It really does look like it returns a dict, and the type system probably agrees.
In these cases it should be fine to fix the documentation, which rightfully is confused about types in a codebase that thus far didn't have any statically declared.

**Review by Byron — APPROVED** [review]
Thanks a lot for yet another fantastic addition of types. 

All changes to the logic where based on the type-system recognising that some fields or variables might be None, so I wouldn't expect any issues arising from this.

**Comment by Byron** [comment]
Please let me know if this is ready for merging, the PR appears to be in draft mode still.

**Comment by Yobmod** [comment]
Now ready for review

File: git/index/base.py (hunk 1 above)
Comment (Yobmod) [reply git/index/base.py]: ok, i changed the docstring then

**Comment by Byron** [comment]
This looks good to me, thanks a lot!

**Pull request merged**